File: scripts/plot_shower_hits_3d.jl
#!/usr/bin/env julia
"""
3-D shower hit visualisation using GLMakie (OpenGL, correct depth rendering).

Shows the terrain mesh, detector box wireframe, and particle hit positions
in local ENU coordinates.  All three axes use the same metre scale so the
scene is physically proportioned.

Usage
-----
  julia --project=. scripts/plot_shower_hits_3d.jl \\
      TERRAIN.h5 BOX.ply PARTICLES.parquet [options]

  Options:
    --output PATH    Output PNG file          [default: shower_hits_3d.png]
    --radius M       Terrain face centroid radius for display (m)  [default: 200]
    --xlim M         Half-width of all three axes (m)              [default: 50]
    --elev DEG       Camera elevation angle (deg)                  [default: 20]
    --azim DEG       Camera azimuth angle, CW from North (deg)     [default: -60]
    --dpi N          Resolution (dots per inch)                    [default: 200]
    --group STR      HDF5 group for terrain vertices/faces
"""

using ArgParse
using GLMakie
using GLMakie.GeometryBasics: Point3f, TriangleFace, Mesh
using HDF5
using LinearAlgebra
using Parquet2
using Statistics
using Tables

const EARTH_RADIUS_M = 6.371e6

# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

function ecef_to_enu(ecef::AbstractMatrix, lat0_deg::Float64, lon0_deg::Float64)
    lat0 = deg2rad(lat0_deg);  lon0 = deg2rad(lon0_deg)
    p0   = EARTH_RADIUS_M .* [cos(lat0)*cos(lon0), cos(lat0)*sin(lon0), sin(lat0)]
    east  = [-sin(lon0),  cos(lon0),  0.0]
    north = [-sin(lat0)*cos(lon0), -sin(lat0)*sin(lon0),  cos(lat0)]
    up    = [ cos(lat0)*cos(lon0),  cos(lat0)*sin(lon0),  sin(lat0)]
    R  = vcat(east', north', up')
    dp = ecef .- p0'
    return (R * dp')'          # (N,3): columns East, North, Up
end

function _ply_type_reader(t::AbstractString)
    t in ("float",  "float32") && return io -> Float64(ltoh(read(io, Float32)))
    t in ("double", "float64") && return io -> ltoh(read(io, Float64))
    t in ("int",    "int32")   && return io -> Int(ltoh(read(io, Int32)))
    t in ("uint",   "uint32")  && return io -> Int(ltoh(read(io, UInt32)))
    t in ("uchar",  "uint8")   && return io -> Int(read(io, UInt8))
    t in ("short",  "int16")   && return io -> Int(ltoh(read(io, Int16)))
    error("Unsupported PLY type: $t")
end

function read_ply(filepath::AbstractString)
    open(filepath, "r") do io
        n_verts = 0; n_faces = 0
        vertex_props = Tuple{String,String}[]
        face_count_type = "uchar"; face_index_type = "uint"
        current = :none
        while true
            line = strip(readline(io))
            line == "end_header" && break
            tokens = split(line); isempty(tokens) && continue
            if tokens[1] == "element"
                current = Symbol(tokens[2])
                tokens[2] == "vertex" && (n_verts = parse(Int, tokens[3]))
                tokens[2] == "face"   && (n_faces = parse(Int, tokens[3]))
            elseif tokens[1] == "property" && current == :vertex && tokens[2] != "list"
                push!(vertex_props, (tokens[end], tokens[2]))
            elseif tokens[1] == "property" && current == :face && tokens[2] == "list"
                face_count_type = tokens[3]; face_index_type = tokens[4]
            end
        end
        xi = findfirst(p -> p[1] == "x", vertex_props)
        yi = findfirst(p -> p[1] == "y", vertex_props)
        zi = findfirst(p -> p[1] == "z", vertex_props)
        readers = [_ply_type_reader(p[2]) for p in vertex_props]
        verts = zeros(Float64, n_verts, 3)
        for i in 1:n_verts
            vals = [r(io) for r in readers]
            verts[i, 1] = vals[xi]; verts[i, 2] = vals[yi]; verts[i, 3] = vals[zi]
        end
        cnt_reader = _ply_type_reader(face_count_type)
        idx_reader = _ply_type_reader(face_index_type)
        faces = Vector{Vector{Int}}(undef, n_faces)
        for j in 1:n_faces
            cnt = cnt_reader(io)
            faces[j] = [idx_reader(io) + 1 for _ in 1:cnt]   # 1-based
        end
        return verts, faces
    end
end

function find_default_group(fid::HDF5.File)::String
    for name in keys(fid)
        g = fid[name]
        isa(g, HDF5.Group) && haskey(g, "vertices") && haskey(g, "faces") && return name
    end
    error("No group with 'vertices' and 'faces' found.")
end

# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------

function parse_args_main()
    s = ArgParseSettings(description="3-D shower hit plot with GLMakie.",
                         add_help=true)
    @add_arg_table! s begin
        "terrain_h5";  help="Terrain HDF5";         arg_type=String; required=true
        "box_ply";     help="Box PLY";               arg_type=String; required=true
        "particles_parquet"; help="Particles parquet"; arg_type=String; required=true
        "--output";    help="Output PNG";            arg_type=String; default="shower_hits_3d.png"
        "--radius";    help="Terrain display radius (m)"; arg_type=Float64; default=200.0
        "--xlim";      help="Half-width of all axes (m)"; arg_type=Float64; default=50.0
        "--elev";      help="Camera elevation (deg)";     arg_type=Float64; default=20.0
        "--azim";      help="Camera azimuth CW from -Y axis (deg)"; arg_type=Float64; default=-60.0
        "--dpi";       help="Resolution";            arg_type=Int; default=200
        "--group";     help="HDF5 group";            arg_type=String; default=""
    end
    return ArgParse.parse_args(s)
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

function main()
    args = parse_args_main()

    # ── Load terrain ──────────────────────────────────────────────────────────
    println("Reading terrain: $(args["terrain_h5"])")
    terrain_ecef, terrain_faces_raw, lat0, lon0 =
        h5open(args["terrain_h5"], "r") do f
            gname = isempty(args["group"]) ? find_default_group(f) : args["group"]
            println("  Group: $gname");  g = f[gname]
            v   = Float64.(read(g["vertices"]))
            fa  = Int.(read(g["faces"]))
            lat = Float64(read(HDF5.attributes(g)["lat_deg"]))
            lon = Float64(read(HDF5.attributes(g)["lon_deg"]))
            v, fa, lat, lon
        end
    println("  $(size(terrain_ecef,1)) vertices, $(size(terrain_faces_raw,1)) faces")

    # ── Load box ──────────────────────────────────────────────────────────────
    println("Reading box: $(args["box_ply"])")
    box_ecef, box_faces_raw = read_ply(args["box_ply"])
    println("  $(size(box_ecef,1)) vertices, $(length(box_faces_raw)) faces")

    # ── Load particles ────────────────────────────────────────────────────────
    println("Reading particles: $(args["particles_parquet"])")
    ds   = Parquet2.Dataset(args["particles_parquet"])
    raw  = Tables.columntable(ds)
    cols = propertynames(raw)
    function getcol(names...)
        for n in names
            sym = Symbol(n)
            sym in cols && return Vector{Float64}(getproperty(raw, sym))
        end
        error("None of $(names) found in columns: $(cols)")
    end
    px_loc = getcol("x", "position_x")
    py_loc = getcol("y", "position_y")
    pz_loc = getcol("z", "position_z")
    ke_raw = getcol("kinetic_energy", "energy", "ekin")
    # unit check: if median > 1e6 assume eV
    ke     = median(ke_raw) > 1e6 ? ke_raw ./ 1e9 : ke_raw
    pdg_sym = findfirst(n -> n in cols, [Symbol("pdg_code"), Symbol("pdg"),
                                          Symbol("particle")])
    pdg = pdg_sym !== nothing ? Vector{Int}(getproperty(raw, cols[pdg_sym])) :
                                fill(0, length(px_loc))
    n_part = length(px_loc)
    println("  $n_part particle hits")

    # ── Convert to ENU ────────────────────────────────────────────────────────
    terrain_enu = ecef_to_enu(terrain_ecef, lat0, lon0)
    box_enu     = ecef_to_enu(box_ecef,     lat0, lon0)
    box_cen     = vec(mean(box_enu, dims=1))
    println("  Box centroid ENU: East=$(round(box_cen[1],digits=1)) m, " *
            "North=$(round(box_cen[2],digits=1)) m, " *
            "Alt=$(round(box_cen[3],digits=1)) m")

    # Particle positions: ECEF offsets from box centroid → rotate to ENU
    lat0r = deg2rad(lat0);  lon0r = deg2rad(lon0)
    east_hat  = [-sin(lon0r),  cos(lon0r),  0.0]
    north_hat = [-sin(lat0r)*cos(lon0r), -sin(lat0r)*sin(lon0r),  cos(lat0r)]
    up_hat    = [ cos(lat0r)*cos(lon0r),  cos(lat0r)*sin(lon0r),  sin(lat0r)]
    R_enu     = vcat(east_hat', north_hat', up_hat')

    part_xyz_off = hcat(px_loc, py_loc, pz_loc)        # (N,3) ECEF offsets
    enu_off      = (R_enu * part_xyz_off')'            # (N,3) ENU offsets
    part_east    = enu_off[:, 1] .+ box_cen[1]
    part_north   = enu_off[:, 2] .+ box_cen[2]
    part_up      = enu_off[:, 3] .+ box_cen[3]

    log_ke = log10.(clamp.(ke, 1e-4, Inf))
    ke_lo  = minimum(log_ke);  ke_hi = maximum(log_ke)

    # ── Filter terrain faces ──────────────────────────────────────────────────
    xl = args["xlim"]
    face_cen_e = [mean(terrain_enu[terrain_faces_raw[i,:] .+ 1, 1]) for i in 1:size(terrain_faces_raw,1)]
    face_cen_n = [mean(terrain_enu[terrain_faces_raw[i,:] .+ 1, 2]) for i in 1:size(terrain_faces_raw,1)]
    nearby = findall(hypot.(face_cen_e, face_cen_n) .<= args["radius"])
    println("  $(length(nearby)) terrain faces within $(args["radius"]) m")

    # ── Interpolate terrain altitude at box centre (ENU origin) ───────────────
    # Find the face with centroid closest to origin, then barycentric interpolation.
    face_dist2 = face_cen_e.^2 .+ face_cen_n.^2
    cf  = argmin(face_dist2)
    ci  = terrain_faces_raw[cf, :] .+ 1          # 1-based vertex indices
    v0  = terrain_enu[ci[1], :]; v1 = terrain_enu[ci[2], :]; v2 = terrain_enu[ci[3], :]
    den = (v1[2]-v2[2])*(v0[1]-v2[1]) + (v2[1]-v1[1])*(v0[2]-v2[2])
    l0  = ((v1[2]-v2[2])*(0-v2[1]) + (v2[1]-v1[1])*(0-v2[2])) / den
    l1  = ((v2[2]-v0[2])*(0-v2[1]) + (v0[1]-v2[1])*(0-v2[2])) / den
    l2  = 1 - l0 - l1
    terrain_alt_origin = l0*v0[3] + l1*v1[3] + l2*v2[3]
    println("  Terrain altitude at box centre (interpolated): $(round(terrain_alt_origin,digits=1)) m")
    box_bottom = minimum(box_enu[:, 3])
    println("  Box bottom: $(round(box_bottom,digits=1)) m  " *
            "(gap = $(round(box_bottom - terrain_alt_origin, digits=1)) m)")

    # Shift box and particles so box bottom sits on the interpolated terrain surface.
    z_shift = terrain_alt_origin - box_bottom
    box_enu[:, 3] .+= z_shift
    box_cen = vec(mean(box_enu, dims=1))
    part_up .+= z_shift

    vis_verts = sort(unique(vcat([terrain_faces_raw[i,:] .+ 1 for i in nearby]...)))
    alt_lo = minimum(terrain_enu[vis_verts, 3])
    alt_hi = maximum(terrain_enu[vis_verts, 3])
    remap  = Dict(v => i for (i,v) in enumerate(vis_verts))

    # ── Build terrain Mesh ────────────────────────────────────────────────────
    mesh_verts = [Point3f(terrain_enu[v,1], terrain_enu[v,2], terrain_enu[v,3])
                  for v in vis_verts]
    mesh_faces = [TriangleFace(remap[terrain_faces_raw[i,1]+1],
                               remap[terrain_faces_raw[i,2]+1],
                               remap[terrain_faces_raw[i,3]+1])
                  for i in nearby]
    terrain_mesh = Mesh(mesh_verts, mesh_faces)
    vert_alts    = Float32[terrain_enu[v, 3] for v in vis_verts]

    # ── Figure ────────────────────────────────────────────────────────────────
    println("Building figure ...")
    GLMakie.activate!(; title="Shower hits 3D")
    set_theme!(theme_latexfonts())

    fig = Figure(size=(1100, 900))

    # Axis3 with equal data-unit scaling
    ax = Axis3(fig[1, 1];
               xlabel = "East  (m)",
               ylabel = "North  (m)",
               zlabel = "Altitude  (m)",
               title  = "3-D shower hits — lat=$(round(lat0,digits=3))°  " *
                         "lon=$(round(lon0,digits=3))°  ($n_part hits)",
               aspect = (1, 1, 1),
               viewmode = :fit,
               limits = (-xl, xl, -xl, xl, terrain_alt_origin - 2.0, terrain_alt_origin + 2*xl))

    # ── Terrain ───────────────────────────────────────────────────────────────
    ms = mesh!(ax, terrain_mesh;
               color      = vert_alts,
               colormap   = :terrain,
               colorrange = (Float32(alt_lo), Float32(alt_hi)),
               alpha      = 0.85,
               transparency = false)
    Colorbar(fig[1, 2]; colormap=:terrain, limits=(alt_lo, alt_hi),
             label="Altitude  (m)", height=Relative(0.7))

    # ── Box wireframe ─────────────────────────────────────────────────────────
    for face in box_faces_raw
        idx = [face; face[1]]
        vx  = box_enu[idx, 1]
        vy  = box_enu[idx, 2]
        vz  = box_enu[idx, 3]
        lines!(ax, vx, vy, vz; color=:red, linewidth=1.0)
    end

    # ── Particle scatter ──────────────────────────────────────────────────────
    pdg_codes  = sort(unique(pdg))
    markers_jl = Dict(22 => :circle, 11 => :rect, -11 => :rect,
                       13 => :utriangle, -13 => :utriangle)
    labels_jl  = Dict(22 => "photon", 11 => "e⁻", -11 => "e⁺",
                       13 => "μ⁻",    -13 => "μ⁺")

    sc = nothing
    for code in pdg_codes
        mask = pdg .== code
        sc = scatter!(ax,
                      part_east[mask], part_north[mask], part_up[mask];
                      color      = log_ke[mask],
                      colormap   = :plasma,
                      colorrange = (ke_lo, ke_hi),
                      markersize = 10,
                      marker     = get(markers_jl, code, :circle),
                      strokewidth = 0.5,
                      strokecolor = :black,
                      label      = get(labels_jl, code, string(code)))
    end
    if sc !== nothing
        axislegend(ax; position=:lt, labelsize=10)
        Colorbar(fig[1, 3]; colormap=:plasma, limits=(ke_lo, ke_hi),
                 label="log₁₀(KE / GeV)", height=Relative(0.7))
    end

    # ── Camera ────────────────────────────────────────────────────────────────
    # Convert elev/azim to Makie's spherical camera angles
    elev_rad = deg2rad(args["elev"])
    azim_rad = deg2rad(args["azim"])
    ax.elevation[] = elev_rad
    ax.azimuth[]   = azim_rad

    # ── Save ─────────────────────────────────────────────────────────────────
    out_dir = dirname(abspath(args["output"]))
    isempty(out_dir) || mkpath(out_dir)
    save(args["output"], fig; px_per_unit = args["dpi"] / 72)
    println("Saved: $(args["output"])")
end

main()

File: scripts/plot_shower_hits_3d.py
#!/usr/bin/env python3
"""
3-D plot of shower particle hits, terrain mesh, and detector box.

Single 3-D axis in local ENU coordinates centred on the detector box,
showing all geometry within --radius metres of the origin.

Usage
-----
  python3 scripts/plot_shower_hits_3d.py TERRAIN.h5 BOX.ply PARTICLES.parquet
      [--output PATH] [--radius M] [--dpi N] [--elev DEG] [--azim DEG]
"""

import argparse
import struct

import h5py
import matplotlib.pyplot as plt
import numpy as np
import pyarrow.parquet as pq
from mpl_toolkits.mplot3d import Axes3D          # noqa: F401 (registers projection)
from mpl_toolkits.mplot3d.art3d import Poly3DCollection

EARTH_RADIUS_M = 6.371e6

# ---------------------------------------------------------------------------
# Coordinate helpers
# ---------------------------------------------------------------------------

def ecef_to_enu(ecef, lat0_deg, lon0_deg):
    lat0 = np.radians(lat0_deg)
    lon0 = np.radians(lon0_deg)
    p0 = EARTH_RADIUS_M * np.array([
        np.cos(lat0)*np.cos(lon0),
        np.cos(lat0)*np.sin(lon0),
        np.sin(lat0)])
    east  = np.array([-np.sin(lon0),  np.cos(lon0),  0.0])
    north = np.array([-np.sin(lat0)*np.cos(lon0), -np.sin(lat0)*np.sin(lon0), np.cos(lat0)])
    up    = np.array([ np.cos(lat0)*np.cos(lon0),  np.cos(lat0)*np.sin(lon0), np.sin(lat0)])
    R  = np.vstack([east, north, up])
    dp = ecef - p0
    return (R @ dp.T).T

# ---------------------------------------------------------------------------
# PLY reader
# ---------------------------------------------------------------------------

def _ply_fmt(t):
    return {'float': ('f', 4), 'float32': ('f', 4),
            'double': ('d', 8), 'float64': ('d', 8),
            'int': ('i', 4), 'int32': ('i', 4),
            'uint': ('I', 4), 'uint32': ('I', 4),
            'uchar': ('B', 1), 'uint8': ('B', 1)}[t]

def read_ply(path):
    with open(path, 'rb') as f:
        n_verts = n_faces = 0
        vprops = []
        fcount_fmt = ('B', 1)
        fidx_fmt   = ('I', 4)
        current = None
        while True:
            line = f.readline().decode('ascii').strip()
            if line == 'end_header':
                break
            tok = line.split()
            if not tok:
                continue
            if tok[0] == 'element':
                current = tok[1]
                if tok[1] == 'vertex': n_verts = int(tok[2])
                if tok[1] == 'face':   n_faces = int(tok[2])
            elif tok[0] == 'property' and current == 'vertex' and tok[1] != 'list':
                vprops.append((tok[-1], tok[1]))
            elif tok[0] == 'property' and current == 'face' and tok[1] == 'list':
                fcount_fmt = _ply_fmt(tok[2])
                fidx_fmt   = _ply_fmt(tok[3])

        xi = next(i for i,(n,_) in enumerate(vprops) if n=='x')
        yi = next(i for i,(n,_) in enumerate(vprops) if n=='y')
        zi = next(i for i,(n,_) in enumerate(vprops) if n=='z')
        fmts = [_ply_fmt(t) for _,t in vprops]
        row_size = sum(s for _,s in fmts)

        verts = np.zeros((n_verts, 3))
        for i in range(n_verts):
            row = f.read(row_size)
            off = 0
            vals = []
            for fmt, size in fmts:
                vals.append(struct.unpack_from('<'+fmt, row, off)[0])
                off += size
            verts[i] = [vals[xi], vals[yi], vals[zi]]

        faces = []
        for _ in range(n_faces):
            cnt = struct.unpack('<'+fcount_fmt[0], f.read(fcount_fmt[1]))[0]
            idx = [struct.unpack('<'+fidx_fmt[0], f.read(fidx_fmt[1]))[0]
                   for _ in range(cnt)]
            faces.append(idx)

    return verts, faces

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('terrain_h5')
    ap.add_argument('box_ply')
    ap.add_argument('particles_parquet')
    ap.add_argument('--output', default='shower_hits_3d.png')
    ap.add_argument('--radius', type=float, default=200.0,
                    help='Terrain face centroid radius for display (m)')
    ap.add_argument('--xlim', type=float, default=50.0,
                    help='Half-width of plot axes (m)')
    ap.add_argument('--dpi',  type=int,   default=200)
    ap.add_argument('--elev', type=float, default=25.0,
                    help='Camera elevation angle (deg)')
    ap.add_argument('--azim', type=float, default=-60.0,
                    help='Camera azimuth angle (deg)')
    ap.add_argument('--group', default='',
                    help='HDF5 group for terrain vertices/faces')
    args = ap.parse_args()

    # ── Load terrain ─────────────────────────────────────────────────────────
    print(f'Reading terrain: {args.terrain_h5}')
    with h5py.File(args.terrain_h5, 'r') as f:
        gname = args.group or next(
            k for k in f if 'vertices' in f[k] and 'faces' in f[k])
        print(f'  Group: {gname}')
        g = f[gname]
        v_raw = g['vertices'][:]
        fa    = g['faces'][:]
        lat0  = float(g.attrs['lat_deg'])
        lon0  = float(g.attrs['lon_deg'])

    terrain_ecef  = v_raw.T if v_raw.shape[0] == 3 else v_raw
    terrain_faces = fa.T    if fa.shape[0]     == 3 else fa
    print(f'  {terrain_ecef.shape[0]} vertices, {terrain_faces.shape[0]} faces')

    # ── Load box ─────────────────────────────────────────────────────────────
    print(f'Reading box: {args.box_ply}')
    box_ecef, box_faces = read_ply(args.box_ply)
    print(f'  {len(box_ecef)} vertices, {len(box_faces)} faces')

    # ── Load particles ────────────────────────────────────────────────────────
    print(f'Reading particles: {args.particles_parquet}')
    tbl    = pq.read_table(args.particles_parquet).to_pydict()
    px_loc = np.array(tbl['x'], dtype=float)   # CORSIKA local x ≈ North
    py_loc = np.array(tbl['y'], dtype=float)   # CORSIKA local y ≈ East
    pz_loc = np.array(tbl['z'], dtype=float)   # CORSIKA local z ≈ Up (relative)
    ke     = np.array(tbl['kinetic_energy'], dtype=float)
    pdg    = np.array(tbl['pdg'], dtype=int)
    n_part = len(px_loc)
    print(f'  {n_part} particle hits')

    # ── Project terrain & box to ENU ─────────────────────────────────────────
    terrain_enu = ecef_to_enu(terrain_ecef, lat0, lon0)
    box_enu     = ecef_to_enu(box_ecef,     lat0, lon0)
    box_cen     = box_enu.mean(axis=0)
    print(f'  Box centroid ENU: East={box_cen[0]:.1f} m, '
          f'North={box_cen[1]:.1f} m, Alt={box_cen[2]:.1f} m')

    # Particle positions are ECEF offsets from the box centroid.
    # Recover ENU via the site rotation matrix.
    lat0_r = np.radians(lat0); lon0_r = np.radians(lon0)
    R_enu = np.vstack([
        np.array([-np.sin(lon0_r),  np.cos(lon0_r),  0.0]),
        np.array([-np.sin(lat0_r)*np.cos(lon0_r),
                   -np.sin(lat0_r)*np.sin(lon0_r), np.cos(lat0_r)]),
        np.array([ np.cos(lat0_r)*np.cos(lon0_r),
                    np.cos(lat0_r)*np.sin(lon0_r), np.sin(lat0_r)]),
    ])
    xyz_offset = np.column_stack([px_loc, py_loc, pz_loc])
    enu_off = (R_enu @ xyz_offset.T).T          # (N,3) ENU offsets from box centroid
    part_east  = enu_off[:, 0] + box_cen[0]
    part_north = enu_off[:, 1] + box_cen[1]
    part_up    = enu_off[:, 2] + box_cen[2]

    # ── Filter terrain faces within radius (centroid-based) ──────────────────
    face_cen_e = terrain_enu[terrain_faces, 0].mean(axis=1)
    face_cen_n = terrain_enu[terrain_faces, 1].mean(axis=1)
    nearby = np.where(np.hypot(face_cen_e, face_cen_n) <= args.radius)[0]
    print(f'  {len(nearby)} terrain faces within {args.radius:.0f} m (centroid)')

    vis_idx = np.unique(terrain_faces[nearby].ravel())
    alt_lo  = terrain_enu[vis_idx, 2].min() if len(vis_idx) else box_cen[2] - 10
    alt_hi  = terrain_enu[vis_idx, 2].max() if len(vis_idx) else box_cen[2] + 10

    # ── Colour scale for kinetic energy ──────────────────────────────────────
    log_ke = np.log10(np.clip(ke, 1e-4, None))
    ke_lo, ke_hi = log_ke.min(), log_ke.max()

    # PDG → marker / label
    markers = {22: 'o', 11: 's', -11: 's', 13: '^', -13: '^'}
    labels  = {22: 'photon', 11: 'e\u207b', -11: 'e\u207a', 13: '\u03bc\u207b', -13: '\u03bc\u207a'}

    # ── Figure ────────────────────────────────────────────────────────────────
    fig = plt.figure(figsize=(9, 8))
    ax  = fig.add_subplot(111, projection='3d')

    n_str = f'{n_part} hits' if n_part else 'no hits'
    ax.set_title(
        f'3-D shower hits  lat={lat0:.3f}\u00b0 lon={lon0:.3f}\u00b0  ({n_str})',
        fontsize=12)
    ax.set_xlabel('East  (m)',     labelpad=8)
    ax.set_ylabel('North  (m)',    labelpad=8)
    ax.set_zlabel('Altitude  (m)', labelpad=8)

    # ── Terrain mesh (plot_trisurf for per-vertex altitude colouring) ────────
    if len(nearby):
        vis_idx   = np.unique(terrain_faces[nearby].ravel())
        remap     = np.empty(terrain_enu.shape[0], dtype=int)
        remap[vis_idx] = np.arange(len(vis_idx))
        local_faces = remap[terrain_faces[nearby]]        # reindexed (M,3)

        tx = terrain_enu[vis_idx, 0]
        ty = terrain_enu[vis_idx, 1]
        tz = terrain_enu[vis_idx, 2]

        ts = ax.plot_trisurf(tx, ty, tz, triangles=local_faces,
                             cmap='terrain', vmin=alt_lo, vmax=alt_hi,
                             alpha=0.70, edgecolor=(0, 0, 0, 0.3), linewidth=0.3)
        fig.colorbar(ts, ax=ax, label='Altitude  (m)',
                     shrink=0.55, pad=0.1, fraction=0.03)

    # ── Box edges ────────────────────────────────────────────────────────────
    for face in box_faces:
        idx = list(face) + [face[0]]
        ax.plot(box_enu[idx, 0], box_enu[idx, 1], box_enu[idx, 2],
                color='red', linewidth=0.7, alpha=0.7)

    # ── Particle hits ─────────────────────────────────────────────────────────
    sc = None
    if n_part > 0:
        for code in np.unique(pdg):
            mask = pdg == code
            sc = ax.scatter(part_east[mask], part_north[mask], part_up[mask],
                            c=log_ke[mask], cmap='plasma', vmin=ke_lo, vmax=ke_hi,
                            s=60, marker=markers.get(code, 'o'),
                            edgecolors='k', linewidths=0.4, zorder=5,
                            depthshade=False,
                            label=labels.get(code, str(code)))
        ax.legend(fontsize=9, loc='upper left', framealpha=0.7)
        cb2 = fig.colorbar(sc, ax=ax, label='log\u2081\u2080(KE / GeV)',
                           shrink=0.55, pad=0.15, fraction=0.03)

    # ── Axis limits ──────────────────────────────────────────────────────────
    r = args.xlim
    ax.set_xlim(-r, r)
    ax.set_ylim(-r, r)
    # z: tight around terrain + a few metres above box top
    box_alt_hi = box_enu[:, 2].max()
    z_lo = alt_lo - 2
    z_hi = max(alt_hi, box_alt_hi) + 5
    ax.set_zlim(z_lo, z_hi)

    ax.view_init(elev=args.elev, azim=args.azim)

    plt.tight_layout()
    plt.savefig(args.output, dpi=args.dpi, bbox_inches='tight')
    print(f'Saved: {args.output}')


if __name__ == '__main__':
    main()
